File: scripts/simplify-geojson.mjs
/**
 * Simplify Bangladesh admin GeoJSON for web rendering.
 *
 * Pipeline:
 *   1. Ramer-Douglas-Peucker vertex reduction  (epsilon = 0.004°, ~400m)
 *   2. Coordinate precision clamp to 4 dp      (~11m, invisible at zoom 5-10)
 *   3. Strip all properties except the ones we use
 *
 * Output sizes (target):
 *   bd-outline.geojson   < 150 KB   (admin0 country polygon)
 *   bd-divisions.geojson < 400 KB   (admin1 division polygons × 8)
 */

import { readFileSync, writeFileSync } from 'fs'
import { resolve, dirname } from 'path'
import { fileURLToPath } from 'url'

const __dir  = dirname(fileURLToPath(import.meta.url))
const ROOT   = resolve(__dir, '..')
const IN_DIR = resolve(ROOT, 'bgd_admin_boundaries.geojson')
const OUT    = resolve(ROOT, 'public', 'data')

// ── Ramer–Douglas–Peucker ─────────────────────────────────────────
function perpendicularDist(pt, lineStart, lineEnd) {
  const dx = lineEnd[0] - lineStart[0]
  const dy = lineEnd[1] - lineStart[1]
  if (dx === 0 && dy === 0) {
    return Math.hypot(pt[0] - lineStart[0], pt[1] - lineStart[1])
  }
  const t = ((pt[0] - lineStart[0]) * dx + (pt[1] - lineStart[1]) * dy) / (dx * dx + dy * dy)
  return Math.hypot(pt[0] - (lineStart[0] + t * dx), pt[1] - (lineStart[1] + t * dy))
}

function rdp(points, epsilon) {
  if (points.length < 3) return points
  let maxDist = 0, maxIdx = 0
  for (let i = 1; i < points.length - 1; i++) {
    const d = perpendicularDist(points[i], points[0], points[points.length - 1])
    if (d > maxDist) { maxDist = d; maxIdx = i }
  }
  if (maxDist > epsilon) {
    const left  = rdp(points.slice(0, maxIdx + 1), epsilon)
    const right = rdp(points.slice(maxIdx),         epsilon)
    return [...left.slice(0, -1), ...right]
  }
  return [points[0], points[points.length - 1]]
}

// ── Coordinate rounding ───────────────────────────────────────────
const P = 4
const r = n => Math.round(n * 10000) / 10000

function roundRing(ring) { return ring.map(([x, y]) => [r(x), r(y)]) }

function simplifyRing(ring, eps) {
  const simplified = rdp(ring, eps)
  // GeoJSON rings must close — ensure first === last
  if (simplified[0][0] !== simplified[simplified.length - 1][0] ||
      simplified[0][1] !== simplified[simplified.length - 1][1]) {
    simplified.push(simplified[0])
  }
  return simplified.map(([x, y]) => [r(x), r(y)])
}

function simplifyGeometry(geom, eps) {
  switch (geom.type) {
    case 'Polygon':
      return { type: 'Polygon', coordinates: geom.coordinates.map(ring => simplifyRing(ring, eps)) }
    case 'MultiPolygon':
      return {
        type: 'MultiPolygon',
        coordinates: geom.coordinates.map(poly => poly.map(ring => simplifyRing(ring, eps))),
      }
    default:
      return geom
  }
}

// ── Process ───────────────────────────────────────────────────────
function processFile(filename, eps, propPicker) {
  const raw  = readFileSync(resolve(IN_DIR, filename), 'utf8')
  const fc   = JSON.parse(raw)
  const out  = {
    type: 'FeatureCollection',
    features: fc.features.map(f => ({
      type: 'Feature',
      properties: propPicker(f.properties),
      geometry: simplifyGeometry(f.geometry, eps),
    })),
  }
  return JSON.stringify(out)
}

// ── admin0: country outline ───────────────────────────────────────
console.log('Processing admin0 (country outline)…')
const outlineJson = processFile('bgd_admin0.geojson', 0.004, () => ({ name: 'Bangladesh' }))
writeFileSync(resolve(OUT, 'bd-outline.geojson'), outlineJson)
console.log('  → bd-outline.geojson', (outlineJson.length / 1024).toFixed(1), 'KB')

// ── admin1: division polygons ─────────────────────────────────────
console.log('Processing admin1 (divisions)…')
const divisionsJson = processFile('bgd_admin1.geojson', 0.004, p => ({
  name:   p.adm1_name,
  pcode:  p.adm1_pcode,
  lat:    p.center_lat,
  lon:    p.center_lon,
}))
writeFileSync(resolve(OUT, 'bd-divisions.geojson'), divisionsJson)
console.log('  → bd-divisions.geojson', (divisionsJson.length / 1024).toFixed(1), 'KB')

// ── admin2: district polygons (64 districts) ──────────────────────
console.log('Processing admin2 (districts)…')
const districtsJson = processFile('bgd_admin2.geojson', 0.003, p => ({
  name:     p.adm2_name,
  division: p.adm1_name,
  pcode:    p.adm2_pcode,
  lat:      p.center_lat,
  lon:      p.center_lon,
}))
writeFileSync(resolve(OUT, 'bd-districts.geojson'), districtsJson)
console.log('  → bd-districts.geojson', (districtsJson.length / 1024).toFixed(1), 'KB')

console.log('Done.')

File: scripts/simplify_geojson.py
"""
Simplify GeoJSON line geometries using Douglas-Peucker algorithm.
Pure Python, no external dependencies.
Usage: python simplify_geojson.py <input.geojson> <output.geojson> [tolerance]
Default tolerance: 0.0002 degrees (~22m)
"""
import json, sys, math

def perp_distance(point, line_start, line_end):
    """Perpendicular distance from point to line segment (XY only)."""
    x0, y0 = point[0], point[1]
    x1, y1 = line_start[0], line_start[1]
    x2, y2 = line_end[0], line_end[1]
    dx, dy = x2 - x1, y2 - y1
    if dx == 0 and dy == 0:
        return math.hypot(x0 - x1, y0 - y1)
    t = ((x0 - x1) * dx + (y0 - y1) * dy) / (dx * dx + dy * dy)
    t = max(0, min(1, t))
    return math.hypot(x0 - (x1 + t * dx), y0 - (y1 + t * dy))

def rdp(coords, tol):
    """Ramer-Douglas-Peucker simplification."""
    if len(coords) < 3:
        return coords
    start, end = coords[0], coords[-1]
    max_dist, max_idx = 0, 0
    for i in range(1, len(coords) - 1):
        d = perp_distance(coords[i], start, end)
        if d > max_dist:
            max_dist, max_idx = d, i
    if max_dist > tol:
        left  = rdp(coords[:max_idx + 1], tol)
        right = rdp(coords[max_idx:], tol)
        return left[:-1] + right
    return [start, end]

def simplify_geometry(geom, tol):
    gtype = geom["type"]
    if gtype == "LineString":
        simplified = rdp(geom["coordinates"], tol)
        if len(simplified) < 2:
            return None
        return {"type": "LineString", "coordinates": simplified}
    if gtype == "MultiLineString":
        lines = [rdp(line, tol) for line in geom["coordinates"]]
        lines = [l for l in lines if len(l) >= 2]
        if not lines:
            return None
        return {"type": "MultiLineString", "coordinates": lines}
    return geom

def round_coords(geom, precision=5):
    """Round coordinate values and drop Z to reduce JSON size."""
    def rnd(c):
        return [round(c[0], precision), round(c[1], precision)]
    gtype = geom["type"]
    if gtype == "LineString":
        return {"type": "LineString", "coordinates": [rnd(c) for c in geom["coordinates"]]}
    if gtype == "MultiLineString":
        return {"type": "MultiLineString", "coordinates": [[rnd(c) for c in line] for line in geom["coordinates"]]}
    return geom

if __name__ == "__main__":
    inp   = sys.argv[1]
    out   = sys.argv[2]
    tol   = float(sys.argv[3]) if len(sys.argv) > 3 else 0.0002

    print(f"Reading {inp}...")
    with open(inp, encoding="utf-8") as f:
        gj = json.load(f)

    orig_count  = len(gj["features"])
    kept        = []
    orig_pts    = 0
    simpl_pts   = 0

    for feat in gj["features"]:
        geom = feat.get("geometry")
        if not geom:
            continue
        # count original points
        if geom["type"] == "LineString":
            orig_pts += len(geom["coordinates"])
        elif geom["type"] == "MultiLineString":
            orig_pts += sum(len(l) for l in geom["coordinates"])

        sgeom = simplify_geometry(geom, tol)
        if sgeom is None:
            continue
        sgeom = round_coords(sgeom, 5)

        if sgeom["type"] == "LineString":
            simpl_pts += len(sgeom["coordinates"])
        elif sgeom["type"] == "MultiLineString":
            simpl_pts += sum(len(l) for l in sgeom["coordinates"])

        kept.append({"type": "Feature", "geometry": sgeom, "properties": feat["properties"]})

    result = {"type": "FeatureCollection", "features": kept}

    print(f"Writing {out}...")
    with open(out, "w", encoding="utf-8") as f:
        json.dump(result, f, separators=(",", ":"))

    import os
    size_mb = os.path.getsize(out) / 1_048_576
    print(f"Features:  {orig_count} → {len(kept)}")
    print(f"Points:    {orig_pts:,} → {simpl_pts:,} ({100*simpl_pts//orig_pts}% kept)")
    print(f"File size: {size_mb:.2f} MB")
